amethyst_gltf/src/format/importer.rs

```diff
@@ -0,0 +1,328 @@
+    } else {
+        import_standard(&data, source, path)
+    }
+}
+
+/// Imports glTF 2.0
+pub fn import<P>(source: Arc<AssetSource>, path: P) -> Result<(Gltf, Buffers), Error>
+where
+    P: AsRef<Path>,
+{
+    import_impl(source, path.as_ref())
+}
+
+fn read_to_end_impl(source: Arc<AssetSource>, path: &Path) -> Result<Vec<u8>, BoxedErr> {
+    Ok(source.load(path.to_str().unwrap())?)
+}
+
+fn read_to_end<P: AsRef<Path>>(source: Arc<AssetSource>, path: P) -> Result<Vec<u8>, BoxedErr> {
+    read_to_end_impl(source, path.as_ref())
+}
+
+fn parse_data_uri(uri: &str) -> Result<Vec<u8>, Error> {
+    let encoded = uri.split(",").nth(1).unwrap();
+    let decoded = base64::decode(&encoded)?;
+    Ok(decoded)
+}
+
+fn load_external_buffers(
+    source: Arc<AssetSource>,
+    base_path: &Path,
+    gltf: &Gltf,
+    mut bin: Option<Vec<u8>>,
+) -> Result<Vec<Vec<u8>>, Error> {
+    let mut buffers = vec![];
+    for (index, buffer) in gltf.buffers().enumerate() {
+        let uri = buffer.uri();
+        let data_res: Result<Vec<u8>, Error> = if uri == "#bin" {
+            Ok(bin.take().unwrap())
+        } else if uri.starts_with("data:") {
+            Ok(parse_data_uri(uri)?)
+        } else {
+            let path = base_path.parent().unwrap_or(Path::new("./")).join(uri);
+            Ok(read_to_end(source.clone(), &path)?)
+        };
+        let data = data_res?;
+
+        if data.len() < buffer.length() {
+            let path = json::Path::new().field("buffers").index(index);
+            return Err(Error::BufferLength(path));
+        }
+        buffers.push(data);
+    }
+    Ok(buffers)
+}
+
+fn validate_standard(unvalidated: gltf::Unvalidated) -> Result<Gltf, Error> {
+    Ok(unvalidated.validate_completely()?)
+}
+
+fn validate_binary(unvalidated: gltf::Unvalidated, has_bin: bool) -> Result<Gltf, Error> {
+    use gltf::json::validation::Error as Reason;
+
+    let mut errs = vec![];
+    {
+        let json = unvalidated.as_json();
+        for (index, buffer) in json.buffers.iter().enumerate() {
+            let path = || json::Path::new().field("buffers").index(index).field("uri");
+            match index {
+                0 if has_bin => if buffer.uri.is_some() {
+                    errs.push((path(), Reason::Missing));
+                },
+                _ if buffer.uri.is_none() => {
+                    errs.push((path(), Reason::Missing));
+                }
+                _ => {}
+            }
+        }
+    }
+
+    if errs.is_empty() {
+        Ok(unvalidated.validate_completely()?)
+    } else {
+        Err(Error::Validation(errs))
+    }
+}
+
+fn import_standard<'a>(
+    data: &'a [u8],
+    source: Arc<AssetSource>,
+    base_path: &Path,
+) -> Result<(Gltf, Buffers), Error> {
+    let unvalidated = Gltf::from_slice(data)?;
+    let gltf = validate_standard(unvalidated)?;
+    let bin = None;
+    let mut buffers = Buffers(vec![]);
+    for buffer in load_external_buffers(source, base_path, &gltf, bin)? {
```

Comment:
Why are you iterating over a vec just to push the elements to another, intially empty one?